feat(lexer): add tokens and lexical analyzer

steguiosaur opened the pull request:
# Description
- `lexer.h`: add lexer (fixes #5) and tokens (fixes #2)
- `lexer.c`: lexer implementations (functions)

## New Keywords
- `define` - define a function
- `maketh` - declare local variable

## Not Implemented

- Noise words

- Increment, Decrement, and others (operators)
- []` symbols

steguiosaur committed:
feat(lexer): add basic lexical analyzer

+ add token types
+ add lexical analyzer
+ print symbol table in main

steguiosaur committed:
refactor(lexer): change lexer function names

+ Show lexemes in symbol table

Issues:
- Cannot print two character tokens
- Cannot detect string literals

steguiosaur committed:
feat(lexer): fix multichar lexemes output

steguiosaur committed:
feat: add other token types, keywords, and symbols

Included several operators, data types, and control flow keywords

steguiosaur committed:
Merge branch 'dev' into lexer

steguiosaur committed:
style: change start_index to curr_i for less text width

+ change start_index to curr_i in lexical analyzer
+ add comments to unclear functionalities

steguiosaur committed:
style: change function order

steguiosaur committed:
Merge branch 'dev' into lexer

steguiosaur committed:
feat: print symbol table on file

steguiosaur committed:
fix(lexer): make lexer not detect an escaped double quote

+ detect double quote escape character in string literals

steguiosaur committed:
fix(main): fix tok memory leak on TK_EOF

+ Add another tokenCleanup to deallocate TK_EOF when loop stops
+ Fix newline in -h option flag

Pilner commented:
Wait til we complete the "majority" of the lexer before merging to the dev branch

steguiosaur committed:
feat: optimize array_textdata allocation on heap

steguiosaur committed:
fix: produce TK_ERR on unterminated TK_STRING

+ Make TK_ERR as token type if string did not terminate up to EOF
+ Fix fileread.c function bracket not aligned on next line

Commit:
feat(lexer): make lexer process comments
+ Implemented:
	- Single Comments
	- Block Comments

Commit:
fix: removed `stdio.h` header
- Removed because it is no longer needed

steguiosaur committed:
Merge pull request #14 from steguiosaur/comments-railey

Implement Lexer Feature: Initial Comments

steguiosaur committed:
feat: ignore comments similar to whitespace

- Remove TK_COMMENTS in TokenType lexer.h
+ Commit fixes #13 which requires to ignore comments in the code
+ Added more test cases for file.rens involving comments

steguiosaur committed:
fix(lexer): fix infinite loop on unterminated block comments

+ Fix infinite loop on unterminated block comment when it reaches EOF
+ Fix block comment terminating on one (#) instead of two hashes (##)

steguiosaur committed:
feat(lexer): add error handler

+ Add error handler for invalid characters and unterminated string
  literals.
+ Create current line number tracker in lexical analyzer used for errors

steguiosaur committed:
fix(main): remove printf for value checking

steguiosaur committed:
docs: add comments for each block of functionality

steguiosaur committed:
feat(lexer): add additional lexemes and token types

Add the following lexemes and token types
[  TK_LBRACKET
]  TK_RBRACKET
%  TK_MODULO
\\ TK_FLOORDIV
++ TK_INCREMENT
-- TK_DECREMENT
+= TK_ASSIGNINC
-= TK_ASSIGNDEC

steguiosaur committed:
fix(lexer): add stdio.h
